geopandas/tests/test_geodataframe_to_file.py

```diff
@@ -0,0 +1,217 @@
+        crs={'init': 'epsg:4326'},
+        geometry=[
+            MultiPolygon((city_hall_boundaries, vauquelin_place)),
+            city_hall_boundaries
+        ]
+    ),
+    # MultiPolygon
+    GeoDataFrame(
+        {'a': [1]},
+        crs={'init': 'epsg:4326'},
+        geometry=[MultiPolygon((city_hall_boundaries, vauquelin_place))]
+    ),
+    # Null geometry and Point
+    GeoDataFrame(
+        {'a': [1, 2]},
+        crs={'init': 'epsg:4326'},
+        geometry=[None, city_hall_entrance]
+    ),
+    # Null geometry and 3D Point
+    GeoDataFrame(
+        {'a': [1, 2]},
+        crs={'init': 'epsg:4326'},
+        geometry=[None, point_3D]
+    ),
+    # Null geometries only
+    GeoDataFrame(
+        {'a': [1, 2]},
+        crs={'init': 'epsg:4326'},
+        geometry=[None, None]
+    )
+
+])
+def geodataframe(request):
+    return request.param
+
+@pytest.fixture(params=[
+    # Points and MultiPoints
+    # It may be a bug in Fiona or in 'ESRI Shapefile' driver:
+    # a gdf with Polygons and MultiPolygons can be written to shapefiles
+    # while a gdf with Points and MultiPoints cannot...
+    GeoDataFrame(
+        {'a': [1, 2]},
+        crs={'init': 'epsg:4326'},
+        geometry=[
+            MultiPoint([city_hall_entrance, city_hall_balcony]),
+            city_hall_balcony
+        ]
+    ),
+    # all shape types
+    GeoDataFrame(
+        {'a': [1, 2, 3, 4, 5, 6]},
+        crs={'init': 'epsg:4326'},
+        geometry=[
+            MultiPolygon((city_hall_boundaries, vauquelin_place)),
+            city_hall_entrance,
+            MultiLineString(city_hall_walls),
+            city_hall_walls[0],
+            MultiPoint([city_hall_entrance, city_hall_balcony]),
+            city_hall_balcony
+        ]
+    ),
+    # all 2D shape types and 3D Point
+    GeoDataFrame(
+        {'a': [1, 2, 3, 4, 5, 6, 7]},
+        crs={'init': 'epsg:4326'},
+        geometry=[
+            MultiPolygon((city_hall_boundaries, vauquelin_place)),
+            city_hall_entrance,
+            MultiLineString(city_hall_walls),
+            city_hall_walls[0],
+            MultiPoint([city_hall_entrance, city_hall_balcony]),
+            city_hall_balcony,
+            point_3D
+        ]
+    )
+])
+def mixed_geom_gdf(request):
+    return request.param
+
+@pytest.fixture(params=['GeoJSON', 'ESRI Shapefile'])
+def ogr_driver(request):
+    return request.param
+
+
+class TestGeoDataFrameToFile():
+
+    def setup_method(self):
+        self.output_dir = tempfile.mkdtemp()
+        self.output_file = os.path.join(self.output_dir, "output_file")
+
+    def teardown_method(self):
+        shutil.rmtree(self.output_dir)
+
+    def test_geodataframe_to_file(self, geodataframe, ogr_driver):
+        self.do_test_geodataframe_to_file(geodataframe, ogr_driver)
+
+    def test_write_gdf_with_mixed_geometries(self, mixed_geom_gdf, ogr_driver):
+        if ogr_driver == 'ESRI Shapefile':
+            with pytest.raises(Exception):
```

Comment:
Can we make this a more explicit exception? Or does it change for fiona below and above 1.8 ?

Reply:
I was actually working on it yesterday. depending on the gdf and the execution environment, I can have 3 different exceptions:

- ValueError: Record's geometry type does not match collection schema's geometry type: 'MultiPoint' != 'Point'
- AttributeError: 'list' object has no attribute 'lstrip'
- RuntimeError: Failed to write record: {'geometry': {'type': 'Point', 'coordinates': (-73.553785, 45.508722)}, 'type': 'Feature', 'id': '1', 'properties': {'a': 2}}
  with a captured log call:
  ERROR    Attempt to write non-polygon (POINT) geometry to POLYGON type shapefile.

I'm trying to figure out how to add more specific exceptions, while keeping the test simple and understandable